include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,130 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// C/C++
+#include <limits>
+
+// deal.II
+#include <deal.II/base/exceptions.h>
+
+// ExaDG
+#include <exadg/postprocessor/time_control.h>
+
+
+TimeControl::TimeControl()
+  : EPSILON(1.0e-10),
+    reset_counter(true),
+    calculate(false),
+    start_time(std::numeric_limits<double>::max()),
+    calculate_every_time_interval(-1.0),
+    calculate_every_time_steps(std::numeric_limits<unsigned int>::max()),
+    counter(0),
+    last_timestep_number_to_evaluate(dealii::numbers::invalid_unsigned_int)
+{
+}
+
+void
+TimeControl::reinit(bool const         calculate_,
+                    double const       start_time_,
+                    double const       calculate_every_time_interval_,
+                    unsigned int const calculate_every_time_steps_)
+{
+  calculate = calculate_;
+
+  if(calculate_)
+  {
+    if(calculate_every_time_steps_ == std::numeric_limits<unsigned int>::max() ||
+       calculate_every_time_steps_ == 0)
+    {
+      AssertThrow(calculate_every_time_interval_ > 0.0 &&
+                    calculate_every_time_interval_ < std::numeric_limits<double>::max(),
+                  dealii::ExcMessage("Input parameters are in conflict."));
+    }
+    else
+    {
+      AssertThrow(calculate_every_time_interval_ < 0.0 ||
+                    calculate_every_time_interval_ >= std::numeric_limits<double>::max(),
+                  dealii::ExcMessage("Input parameters are in conflict."));
+    }
+  }
+
+  start_time                    = start_time_;
+  calculate_every_time_interval = calculate_every_time_interval_;
+  calculate_every_time_steps    = calculate_every_time_steps_;
+}
+
+// If reinit() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, unsigned int const time_step_number) const
+{
+  if(calculate == false)
+    return false;
+
+  bool evaluate = false;
+
+  if(calculate_every_time_steps > 0)
+  {
+    if(time > start_time && (time_step_number - 1) % calculate_every_time_steps == 0)
+    {
+      evaluate = true;
+    }
+  }
+  else if(calculate_every_time_interval > 0.0)
+  {
+    // The current time might be larger than start_time. In that case, we first have to
+    // reset the counter in order to avoid that output is written every time step.
+    if(reset_counter)
+    {
+      counter += int((time - start_time + EPSILON) / calculate_every_time_interval);
+      reset_counter = false;
+    }
+
+    if((time > (start_time + counter * calculate_every_time_interval - EPSILON)))
+    {
+      evaluate = true;
+      ++counter;
```

Comment:
Question: should we do this manually? There is no guarantee that evaluate is called after it is checked if evaluation is needed.

Reply:
Do you mean incrementing the counter? What is the solution that you have in mind?

Reply:
Principally, the postprocessing routine could take care of counter. However, it is then more difficult the fix the problem with the restart, and all the postprocessing routines need a counter. So I would accept the currrent solution, maybe with a comment to make users aware of this subtlety.

Reply:
A different option would be a member function ```increment_counter()```
```
time_control.increment_counter();
```
 or ```operator++```.
```
++time_control
```

Reply:
Yes, but I think this variant can also lead to errors. Think about implementing a new postprocessing routine but forgetting to increment `TimeControl`` at  the end of the `evaluate()` function. Then, the postprocessing will be invoked after every time step and the parameters (time step interval and time interval) are just ignored. Since counter is not incremented, all output files have the same output number/counter in the filename (but output is written every timestep!). For the current variant, I would say it is less likely to forget calling `evaluate()` after the if-statement. What do you think? I would personally prefer the variant that does require one line of code for every postprocessing routine, given that both variants can be misused.

Reply:
OK. I think I will keep it as it is (the increment in TimeControl) :)